src/OrbitQt/AnnotatingSourceCodeDialog.h

```diff
@@ -0,0 +1,106 @@
+// destructed when the dialog gets closed. Use the free function `OpenAndDeleteOnClose` to
+// obtain this behaviour.
+//
+// Use `QDialog::exec` if you require a modal dialog which blocks the code path execution while the
+// dialog is shown.
+class AnnotatingSourceCodeDialog : public orbit_code_viewer::Dialog {
+  Q_OBJECT
+
+ public:
+  using orbit_code_viewer::Dialog::Dialog;
+
+  // Same interface as OrbitApp::RetrieveModuleWithDebugInfo;
+  using RetrieveModuleWithDebugInfoCallback =
+      std::function<orbit_base::Future<ErrorMessageOr<std::filesystem::path>>(const std::string&,
+                                                                              const std::string&)>;
+
+  void SetDisassemblyCodeReport(orbit_code_report::DisassemblyReport report) {
+    report_ = std::move(report);
+  }
+
+  // Enable the heatmap side bar on the left. This requires a disassembly code report to be set with
+  // `SetDisassemblyCodeReport`.
+  void EnableHeatmap(orbit_code_viewer::FontSizeInEm heatmap_bar_width);
+
+  // Call this function to trigger the annotation process. It requires a disassembly code report to
+  // be set before hand by calling `SetDisassemblyCodeReport`.
+  void AddAnnotatingSourceCode(orbit_client_protos::FunctionInfo function_info,
+                               RetrieveModuleWithDebugInfoCallback callback);
+
+ signals:
+  void SourceCodeAvailable();
+  void SourceCodeLoaded();
+
+ private:
+  void DialogActionButtonClicked();
+  [[nodiscard]] bool LoadElfFile(const std::filesystem::path& local_file_path);
+  [[nodiscard]] bool LoadLocationInformationFromElf();
+  [[nodiscard]] bool DetermineLocalSourceFilePath();
+  void LoadSourceCode();
+  void HandleDebugInfo(const ErrorMessageOr<std::filesystem::path>& local_file_path);
+  void ChooseFile();
+  void HandleSourceCode(const QString& source_file_contents);
+  void HandleAnnotations();
+
+  orbit_client_protos::FunctionInfo function_info_;
+  std::optional<orbit_code_report::DisassemblyReport> report_;
+  RetrieveModuleWithDebugInfoCallback retrieve_module_with_debug_info_;
+
+  std::chrono::steady_clock::time_point starting_time_ = std::chrono::steady_clock::now();
+  std::shared_ptr<MainThreadExecutor> main_thread_executor_ =
+      orbit_qt_utils::MainThreadExecutorImpl::Create();
+
+  enum class ButtonAction { kNone, kChooseFile, kAddAnnotations, kHide };
+  ButtonAction awaited_button_action_ = ButtonAction::kNone;
+
+  std::filesystem::path local_source_file_path_;
+  std::unique_ptr<orbit_object_utils::ElfFile> elf_file_;
+  orbit_grpc_protos::LineInfo location_info_;
+  std::vector<orbit_code_report::AnnotatingLine> annotations_;
+};
+
+// This function opens the given dialog and ensures it is deleted when closed.
+// Note, this function returns immediately after opening the dialog, NOT when it is closed. Use
+// `QDialog::exec` to wait for the dialog.
+QPointer<AnnotatingSourceCodeDialog> OpenAndDeleteOnClose(
```

Comment:
Nit and not a strong opinion: This could be a static member of `AnnotatingSourceCodeDialog` since it's tightly coupled to that scope.